### DISCUSSION TITLE
Langraph- how to use checkpointer

### DISCUSSION THREAD

**Discussion by manju246** [community/community#143484] [Programming Help]
### Select Topic Area

Question

### Body

I am trying to add memory in my RAG using Langgraph. I have added memory like

checkpointer = MemorySaver()

app = workflow.compile(checkpointer = checkpointer)

I am getting below error:
Type error : object of type QueryCaptionResult is not serializable.

**Comment by LiteBrite82** [comment]
Thanks for posting in the GitHub Community, @manju246 ! 

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.

**Comment by michaelaltao** [comment]
Hi @manju246 
Hope you are doing well.

Here are some of my solutions to solve this problem.

The error TypeError: object of type QueryCaptionResult is not serializable suggests that the QueryCaptionResult object, used somewhere in your workflow, cannot be serialized by the default pickle or json serialization in LangGraph.

To fix this issue, you need to either:
Implement Custom Serialization for QueryCaptionResult: Create custom serialization (and deserialization) functions for QueryCaptionResult so it can be stored in memory properly.
Exclude Non-Serializable Objects from Memory: Modify the code to avoid storing QueryCaptionResult in memory or transform it into a serializable type (e.g., dictionary, string).

Here are a few solutions to address this:
Solution 1: Custom Serialization for QueryCaptionResult
If you have control over QueryCaptionResult, implement a custom method to convert it to a serializable form (like a dictionary or JSON string). You can also create a helper function to handle the serialization.


```import json

class QueryCaptionResult:
    def __init__(self, data):  # Example init
        self.data = data  # Adjust as per actual attributes

    # Custom serialization
    def to_dict(self):
        return {"data": self.data}  # Convert to dictionary

    @classmethod
    def from_dict(cls, dict_data):
        return cls(data=dict_data["data"])

# Then use these methods before saving to memory:
serializable_data = QueryCaptionResult.to_dict()
# For deserialization, you could use QueryCaptionResult.from_dict(serializable_data)
```

Update the code to convert QueryCaptionResult objects into a serializable format before storing them in MemorySaver.

Solution 2: Exclude Non-Serializable Data from Memory
If QueryCaptionResult data isn't necessary for memory, you can alter your workflow to avoid saving this particular data type in memory.

```# In your workflow or a custom memory-saving function:
if isinstance(data, QueryCaptionResult):
    # Either skip or convert to a more primitive type
    data = str(data)  # Or use any other approach suitable for your case
```

Solution 3: Use a Serializer Library
If QueryCaptionResult is a complex or nested data structure, you can leverage a library like dill or cloudpickle, which supports serializing a broader range of objects.

```
import dill as pickle  # or import cloudpickle as pickle

# Saving
with open("query_caption_result.pkl", "wb") as f:
    pickle.dump(QueryCaptionResult, f)

# Loading
with open("query_caption_result.pkl", "rb") as f:
    loaded_data = pickle.load(f)
```

Replace the default serialization with dill or cloudpickle when you set up the memory-saving functionality.

Hope to be helpful for you!
Please mark up this answer as done if then.

Thank you

**Reply by manju246** [reply]
Thank you @collinpksmith. I will definitely try this.
